File: integration-known-function.c

#include <stdio.h>

float func(float x){
    return 5*(x*x*x) - 3*(x*x) + 2*x + 1;
}

float absVal(float x){
    return x>=0? x : x*-1;
}

float relErr(float x1, float x2){
    return absVal((x1-x2)/x1);
}

int power(int base, int exp){
    int ans = 1;
    for(int i=0;i<exp;i++) ans*=base;
    return ans;
}

int main() {
    float x1, x2, h;
    float e = 0.001;
    int k = 3,p = 1;
    printf("Please enter the range of integral: ");
    printf("\nEnter x1: ");
    scanf("%f", &x1);
    printf("\nEnter x2: ");
    scanf("%f", &x2);
    h = (x2-x1)/2;
    
    float integral = 0;
    float x = x1;
    for(int i=1;i<=k;i++){
        //printf("\nDebugging: x:%f f(x):%f",x,func(x));
        if(i==1 || i==k) integral += (func(x)/2);
        else integral += func(x);
        x = x+h;
    }
    integral *= h;
    printf("\n The integral so obatined is: %f", integral);
    
    float oldIntegral = integral;
    
    do{
        h = h/2;
        k = power(2,++p) +1;
        oldIntegral = integral;
        integral = 0;
        x = x1;
        printf("\n=========================\n");
        printf("Starting calc with p=%d x1=%f",p, x1);
        
        for(int i=1;i<=k;i++){
            printf("\nx=%f h=%f k=%d", h, x, k);
            if(i==1 || i==k){
                integral += (func(x)/2);
            }
            else{
                integral += (func(x));
            }
        }
        integral *= h;
    }while(relErr(integral, oldIntegral)>e);
    
    printf("\nThe value of integral obtained after improvisation is: %f with k=%d and p=%d", integral, k, p);
    return 0;
}

File: integration-known-function.java
import java.util.Scanner;

class integration{

    static double relErr(double x1, double x2){
        return Math.abs((x2-x1));
    }

    static double function(double x){
        //return num; //Starting with our linear function here
        return 5*(x*x*x) - 3*(x*x) + 2*x + 1;
    }

    public static void main(String[]args){
        final double e = 0.0001; //Accepted Value of Relative Error

        Scanner myScanner = new Scanner(System.in);
        System.out.println("Please enter the range of integration: ");
        double x1 = myScanner.nextInt();
        double x2 = myScanner.nextInt();
        double h = (x2-x1)/2;
        int k = 3;
        double x = x1; //Variable to temporarily store the initial x
        double integral = 0; //Stores the value of the integral initially calculated
        for(int i=1;i<=k;i++){
            if(i==1 || i==k) integral += (function(x)/2);
            else integral += function(x);
            x = x+h;

        }
        integral *= h;
        System.out.println("The initial value of the integral so calculated is: " + integral);

        double oldIntegral = integral+1.0;

        while(relErr(integral, oldIntegral)>e){
            h = h/2;
            x = x1;
            oldIntegral = integral;
            integral = 0;
            for(;x<=x2;x=x+h){
                if(x==x1 || x==x2) integral += (function(x)/2);
                integral += function(x);
                x = x+h;
            }
            integral *= h;
            System.out.printf("\nThe value of the integral obtained here is: %.4f", integral);
            System.out.printf("\nThe relative error between the two integrals is : %.5f",relErr(integral, oldIntegral));
        }
    }
}